### PR TITLE
New rule S1128 Unused import check

### LINKED ISSUE #643: Rule: Unnecessary imports should be removed
Implements [RSPEC-1128](https://jira.sonarsource.com/browse/RSPEC-1128)

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #643

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnusedImportCheck.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  public void visitImportClause(ImportClauseTree tree) {
+    tree.descendants()
+      .filter(t -> t.is(Tree.Kind.BINDING_IDENTIFIER))
+      .map(t -> (IdentifierTree) t)
+      .forEach(identifierTree -> identifierTree.symbol().ifPresent(importedSymbols::add));
+    super.visitImportClause(tree);
+  }
+
+  @Override
+  public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnusedImportCheck.java]: Code Smell Code Smell: Remove this method to simply inherit it. ([squid:S1185](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1185))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.javascript%3Ajavascript&branch=issue-643&issues=AWFr5GtOUBMQ29P18y6N&open=AWFr5GtOUBMQ29P18y6N)

[ISSUE_KEY]: AWFr5GtOUBMQ29P18y6N

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S1128.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "2min"
+  },
+  "tags": [
+    "es2015",
+    "unused"
+  ],
+  "defaultSeverity": "Minor",
+  "ruleSpecification": "RSPEC-1128",
+  "sqKey": "UselessImportCheck"
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S1128.json]: @saberduck something is wrong here:)

File: javascript-checks/src/test/resources/checks/unusedImport.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// @flow
+
+import type {UserID, //Noncompliant {{Remove this unused import of 'UserID'.}}
+    User} from "MyTypes"; // Noncompliant {{Remove this unused import of 'User'.}}
+
+import A from 'a';      // Noncompliant {{Remove this unused import of 'A'.}} ,A isn't used
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/unusedImport.js]: @saberduck `,A isn't used`?

File: javascript-checks/src/test/resources/checks/unusedImport.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import { B1 } from 'b';
+
+console.log("My first JavaScript...");
+
+import { B1 } from 'b'; // Noncompliant {{'B1' is already imported; remove this redundant import.}}
+
+console.log(B1);
+
+import * as test from 'starimport'; // Noncompliant
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/unusedImport.js]: @saberduck can we also test `import {a as b} from 'mymodule'`?

File: its/ruling/src/test/expected/javascript-S1128.json [hunk 5]
```diff
@@ -0,0 +1,468 @@
+{
+'project:babylon/src/parser/expression.js':[
+24,
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/javascript-S1128.json]: @saberduck FPs (I've simpified pattern):

```
import * as usedStarImport from 'usedstarimport';
var x: usedStarImport.foo = 42;

```

**Comment by vilchik-elena** [comment]
@saberduck exclude "React"

File: javascript-checks/src/test/resources/checks/ReferenceError.js [hunk 6]
```diff
@@ -51,3 +51,28 @@ function f1() {
...
+type OtherType = number;
+
+function blah<T>(x: T): T  {
+
+}
+
+// these flow constructs are not yet supported and they don't create correct symbols
+
+class FlowFunctionType {
+  f: <T>(x: T) => T;  // Noncompliant, FP
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/ReferenceError.js]: @saberduck could you add precise issue location to these FPs?

File: sonar-javascript-plugin/pom.xml [hunk 7]
```diff
@@ -116,7 +116,7 @@
               <rules>
                 <requireFilesSize>
                   <minsize>3300000</minsize>
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/pom.xml]: @saberduck increase minsize

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java [hunk 8]
```diff
@@ -231,10 +237,18 @@ public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
     leaveScope();
   }
 
+  private void declareGenericFlowParameters(@Nullable FlowGenericParameterClauseTree flowGenericParameterClauseTree) {
+    if (flowGenericParameterClauseTree != null) {
+      List<IdentifierTree> genericParams = flowGenericParameterClauseTree.genericParameters().stream().map(FlowGenericParameterTree::identifier).collect(Collectors.toList());
+      declareParameters(genericParams);
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java]: @saberduck I'm not sure if it's good idea to use Kind.Parameter for Generic parameter. Should we create a new kind? Let's discuss it

File: javascript-frontend/src/main/resources/org/sonar/javascript/tree/symbols/globals.json [hunk 9]
```diff
@@ -1190,5 +1190,36 @@
     "GM_setValue",
     "GM_xmlhttpRequest",
     "unsafeWindow"
+  ],
+  "flow": [
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/resources/org/sonar/javascript/tree/symbols/globals.json]: @saberduck we should update docs https://github.com/SonarSource/SonarJS/blob/master/docs/DOC.md#advanced-configuration (add flow)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java (hunk 8 above)
Comment (vilchik-elena) [reply javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java]: another reason to have another kind:
image

I think it should be another rule for generic parameters (or at least another message)

**Comment by vilchik-elena** [comment]
@saberduck I think it will be good to extend unused parameter rule test cases (as there are new issues). 

And btw weird issues:

image

why only one issue for `T`?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnusedImportCheck.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+  @Override
+  public void visitImportDeclaration(ImportDeclarationTree tree) {
+    if (isReactImport(tree)) {
+      return;
+    }
+    super.visitImportDeclaration(tree);
+  }
+
+  private boolean isReactImport(ImportDeclarationTree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnusedImportCheck.java]: Code Smell Code Smell: Make "isReactImport" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.javascript%3Ajavascript&branch=issue-643&issues=AWFwj7P-h1v1MSKKr6Yh&open=AWFwj7P-h1v1MSKKr6Yh)

[ISSUE_KEY]: AWFwj7P-h1v1MSKKr6Yh

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java (hunk 8 above)
Comment (saberduck) [reply javascript-frontend/src/main/java/org/sonar/javascript/tree/symbols/HoistedSymbolVisitor.java]: As discussed, indeed there is another rule https://jira.sonarsource.com/browse/RSPEC-2326 for unused type parameters. Created #927 to implement it

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S1128.html [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+<h2>Compliant Solution</h2>
+<pre>
+import { B1 } from 'b';
+
+console.log("My first JavaScript...");
+
+console.log(B1);
+</pre>
+<h2>Exceptions</h2>
+<p>Imports for types mentioned in comments, such as Javadocs, are ignored.</p>
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S1128.html]: @saberduck we need to remove this exception

**Comment by vilchik-elena** [comment]
Fix rule description and feel free to merge

**Pull request merged**